Adds support for coordinate subscripting.

Raahul-Singh opened the pull request:
<!-- Please fill below the issue number which this code fixes -->
Fixes #366

<!-- Please tag any reviewer here -->

pep8speaks commented:
Hello @Raahul-Singh! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`src/einsteinpy/coordinates/core.py`](https://github.com/einsteinpy/einsteinpy/blob/a05750e37e1829b18885155982d98e98b63e4251/src/einsteinpy/coordinates/core.py):

> [Line 33:32](https://github.com/einsteinpy/einsteinpy/blob/a05750e37e1829b18885155982d98e98b63e4251/src/einsteinpy/coordinates/core.py#L33): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 33:44](https://github.com/einsteinpy/einsteinpy/blob/a05750e37e1829b18885155982d98e98b63e4251/src/einsteinpy/coordinates/core.py#L33): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 33:56](https://github.com/einsteinpy/einsteinpy/blob/a05750e37e1829b18885155982d98e98b63e4251/src/einsteinpy/coordinates/core.py#L33): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 33:73](https://github.com/einsteinpy/einsteinpy/blob/a05750e37e1829b18885155982d98e98b63e4251/src/einsteinpy/coordinates/core.py#L33): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 40:1](https://github.com/einsteinpy/einsteinpy/blob/a05750e37e1829b18885155982d98e98b63e4251/src/einsteinpy/coordinates/core.py#L40): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 41:25](https://github.com/einsteinpy/einsteinpy/blob/a05750e37e1829b18885155982d98e98b63e4251/src/einsteinpy/coordinates/core.py#L41): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 43:1](https://github.com/einsteinpy/einsteinpy/blob/a05750e37e1829b18885155982d98e98b63e4251/src/einsteinpy/coordinates/core.py#L43): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 145:32](https://github.com/einsteinpy/einsteinpy/blob/a05750e37e1829b18885155982d98e98b63e4251/src/einsteinpy/coordinates/core.py#L145): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 145:48](https://github.com/einsteinpy/einsteinpy/blob/a05750e37e1829b18885155982d98e98b63e4251/src/einsteinpy/coordinates/core.py#L145): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 145:66](https://github.com/einsteinpy/einsteinpy/blob/a05750e37e1829b18885155982d98e98b63e4251/src/einsteinpy/coordinates/core.py#L145): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 145:85](https://github.com/einsteinpy/einsteinpy/blob/a05750e37e1829b18885155982d98e98b63e4251/src/einsteinpy/coordinates/core.py#L145): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 155:25](https://github.com/einsteinpy/einsteinpy/blob/a05750e37e1829b18885155982d98e98b63e4251/src/einsteinpy/coordinates/core.py#L155): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 157:1](https://github.com/einsteinpy/einsteinpy/blob/a05750e37e1829b18885155982d98e98b63e4251/src/einsteinpy/coordinates/core.py#L157): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace
> [Line 228:32](https://github.com/einsteinpy/einsteinpy/blob/a05750e37e1829b18885155982d98e98b63e4251/src/einsteinpy/coordinates/core.py#L228): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 228:48](https://github.com/einsteinpy/einsteinpy/blob/a05750e37e1829b18885155982d98e98b63e4251/src/einsteinpy/coordinates/core.py#L228): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 228:66](https://github.com/einsteinpy/einsteinpy/blob/a05750e37e1829b18885155982d98e98b63e4251/src/einsteinpy/coordinates/core.py#L228): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 228:80](https://github.com/einsteinpy/einsteinpy/blob/a05750e37e1829b18885155982d98e98b63e4251/src/einsteinpy/coordinates/core.py#L228): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 228:95](https://github.com/einsteinpy/einsteinpy/blob/a05750e37e1829b18885155982d98e98b63e4251/src/einsteinpy/coordinates/core.py#L228): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ':'
> [Line 238:25](https://github.com/einsteinpy/einsteinpy/blob/a05750e37e1829b18885155982d98e98b63e4251/src/einsteinpy/coordinates/core.py#L238): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 239:38](https://github.com/einsteinpy/einsteinpy/blob/a05750e37e1829b18885155982d98e98b63e4251/src/einsteinpy/coordinates/core.py#L239): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

shreyasbapat commented:
Hey @Raahul-Singh you need to take care of code linting. 

You will have to run `tox -e reformat` to fix this :)

Raahul-Singh commented:
> Hey @Raahul-Singh you need to take care of code linting.
> 
> You will have to run `tox -e reformat` to fix this :)

And with one command you have helped me solve most of my pep8 issues. Thanks for this!!

shreyasbapat commented on src/einsteinpy/coordinates/core.py:
```diff
@@ -30,13 +30,22 @@ def __init__(self, x, y, z):
         self.z = z
         super().__init__(x.si.value, y.si.value, z.si.value)
         self.system = "Cartesian"
+        self.coordinates = {
```
I guess the self.coordinates is a vague way of indicating what's already coordinate!

Raahul-Singh replied:
Should I go with `self.data` ?

shreyasbapat commented on src/einsteinpy/coordinates/core.py:
```diff
@@ -30,13 +30,22 @@ def __init__(self, x, y, z):
         return self.__repr__()
 
+    def __getitem__(self, item):
+        return self.coordinates[item]
```
Same here

shreyasbapat commented:
Also, this would need a rebase

Raahul-Singh commented:
> Can we also write unit tests for this, that would make the PR mergeable, along with the changes O proposed.
> Can we do `self._dimension` instead ?

Where should I put the test? Which file?
I'm a little confused as this is a simple `__getitem__` implementation.

shreyasbapat commented:
Sure. You can put tests in this file: 

https://github.com/einsteinpy/einsteinpy/blob/master/src/einsteinpy/tests/test_coordinates/test_coord_transform.py

shreyasbapat commented:
Still, we need to be sure that it works always! :)

ritzvik commented:
My two cents for this PR.
Can we make `obj[0]`, `obj[1]` and `obj[2]` the aliases for `obj["x"]`, `obj["y"]` and `obj["z"]` ? Same for `r, theta, phi`. 

Anyway, @Raahul-Singh , wait for @shreyasbapat go-ahead before implementing this.

shreyasbapat commented:
My end goal with the issue was the same. But then it came to my mind, how do we know which one is 0,1 and 2. If you feel this is not a big issue @ritzvik , and have some solution. I am more inclined to the solution you suggested.

ritzvik commented:
I guess it's pretty evident in Cartesian case, `0,1,2` for `x,y,z`.
`r, theta, phi` is the order used in [wikipedia](https://en.wikipedia.org/wiki/Spherical_coordinate_system), so we can go with that in case of Spherical and BL.

We can add these entries into the same `self.coordinates` as coded by @Raahul-Singh , where `0,1,2` can be the keys and `self.x, self.y, self.z` can be the values. What do you say @shreyasbapat ?

Also, @Raahul-Singh , please change the same to another other than `self.coordinates`. It seems very ambiguous

shreyasbapat commented:
If you feel that's obvious, I am fine. 
Rename it to `self._dimensions`

shreyasbapat commented:
This can't show up in the docs. We need to make it private

Raahul-Singh commented:
Hey! I've changed `self.coordinates` to `self._dimension` . I've also added a test. The simplest way to make this subscript-able by index is to use a tuple rather than a dictionary. Tuple instead of a list because list is mutable. However, `self._dimension[-1]` would get the name of the coordinate whereas all others would get quantities. Would this not cause occasional errors, if the user is not familiar with it?

ritzvik commented on src/einsteinpy/coordinates/core.py:
```diff
@@ -30,13 +30,22 @@ def __init__(self, x, y, z):
+            "z": self.z,
+            "system": "Cartesian",
+        }
 
```
Add a line here,
```suggestion
        self._dimension_order = ("x", "y", "z")
```

ritzvik commented on src/einsteinpy/coordinates/core.py:
```diff
@@ -30,13 +30,22 @@ def __init__(self, x, y, z):
         return self.__repr__()
 
+    def __getitem__(self, item):
+        return self._dimension[item]
```
And a change here,
```suggestion
        if isinstance(item, (int, np.integer)):
            return self._dimension[self._dimension_order[item]]
        return self._dimension[item]
```

Does this two comments address this https://github.com/einsteinpy/einsteinpy/pull/387#issuecomment-578191030 ?

ritzvik replied:
Also, please write a docstring for `__getitem__`s. Users should know that objects are subscriptable.

shreyasbapat commented:
> Hey! I've changed `self.coordinates` to `self._dimension` . I've also added a test. The simplest way to make this subscript-able by index is to use a tuple rather than a dictionary. Tuple instead of a list because list is mutable. However, `self._dimension[-1]` would get the name of the coordinate whereas all others would get quantities. Would this not cause occasional errors, if the user is not familiar with it?

I guess we will document that somewhere. Put a comment in the code for the same. I will get it sorted in my next PR. And please fix the sggestions by @ritzvik . And rebase and squash!

shreyasbapat commented:
@Raahul-Singh are you still on it?

Raahul-Singh commented:
> @Raahul-Singh are you still on it?

Oh, yes. I'll get it done by tonight. Just a busy couple of days. 😅 I'm sorry.

Raahul-Singh commented:
Does this work?

ritzvik commented:
> Does this work?

Yups! Seems good enough!
BTW, maybe, you rebased the branch wrong. The PR is showing "25 files changed".

Raahul-Singh commented:
> > Does this work?
> 
> Yups! Seems good enough!
> BTW, maybe, you rebased the branch wrong. The PR is showing "25 files changed".

Sorry. :sweat_smile: I am not that experienced with rebasing yet. Could you help me with fixing it?

ritzvik commented on src/einsteinpy/coordinates/core.py:
```diff
@@ -217,6 +263,14 @@ def __init__(self, r, theta, phi, a):
+            "a": self.a,
+            "system": "BoyerLindquist",
+        }
+        self._dimension_order = ("r", "theta", "phi", "a")
```
"a" is not a dimension here. It is an attribute of space-time. Can you remove it from `_dimension_order`?

ritzvik replied:
P.S. : feature of the space spanned by BL coordinates

ritzvik replied:
This one thing, and then we can rebase!

ritzvik commented:
I guess you somehow included all the previous commits in 
21d9a97 .

So I would say, go to the project root and do this:
```
git reset --hard HEAD~3 # makes a hard reset to 3 commits behind
git checkout dfe2eac -- ./src/einsteinpy/coordinates/core.py # copies the file from your latest commit
```
You would find `core.py` file staged and make a new commit. After this, rebase with upstream
```
git pull --rebase upstream master
```

I believe the changes in test file are pretty small. Just copy-paste them and create another commit.

That's it

Raahul-Singh commented:
Thanks @ritzvik :smile:

ritzvik commented:
@shreyasbapat LGTM.
If you are satisfied, merge it.

shreyasbapat commented on src/einsteinpy/coordinates/core.py:
```diff
@@ -30,13 +30,36 @@ def __init__(self, x, y, z):
+                Name of the parameter or it's index.
+                If `system`, Name of coordinate is returned.
+        """
+        if isinstance(item, (int, np.integer)):
```
Very good @Raahul-Singh ! This is exactlt what was expected!

Raahul-Singh replied:
Thank you for helping me along the way!
